File: WorkBranch/backend/service/session_service/session.py
import asyncio
from typing import List, Optional, Dict, Any, Callable, Awaitable

from singleton import get_session_history, get_conversation_service, get_conversation_dao, get_conversation_buffer
from service.user_service.session_history import SessionHistory
from service.session_service.conversation_service import ConversationService
from service.session_service.conversation_buffer import ConversationBuffer
from data.conversation_dao import ConversationDAO, Session, Conversation


class SessionService:
    _instance = None
    _initialized = False

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __init__(self):
        if SessionService._initialized:
            return
        SessionService._initialized = True

        self._session_history: SessionHistory = get_session_history()
        self._conversation_service: ConversationService = get_conversation_service()
        self._conversation_buffer: ConversationBuffer = get_conversation_buffer()
        self._dao: ConversationDAO = get_conversation_dao()
        self._lock = asyncio.Lock()

    def create_session(self, title: str = "新会话") -> Session:
        return self._session_history.create_session(title)

    def delete_session(self, session_id: int) -> bool:
        conversations = self._dao.list_conversations_by_session(session_id)

        async def _async_delete():
            for conversation in conversations:
                await self._conversation_service.delete_conversation(conversation.id)

        try:
            loop = asyncio.get_event_loop()
            if loop.is_running():
                asyncio.create_task(_async_delete())
            else:
                loop.run_until_complete(_async_delete())
        except RuntimeError:
            pass

        self._session_history.delete_session(session_id)
        return True

    def list_sessions(self) -> List[Session]:
        return self._session_history.list_sessions()

    def get_session(self, session_id: int) -> Optional[Session]:
        return self._session_history.get_session(session_id)

    async def create_conversation(
        self,
        session_id: int,
        workspace_id: Optional[str] = None,
        parent_conversation_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        session = self.get_session(session_id)
        if not session:
            raise ValueError(f"Session {session_id} not found")

        if parent_conversation_id is not None:
            parent_conversation = self._dao.get_conversation_by_id(parent_conversation_id)
            if not parent_conversation:
                raise ValueError(f"Conversation {parent_conversation_id} not found")
            if parent_conversation.session_id != session_id:
                raise ValueError("Parent conversation does not belong to this session")

        conversation_id = await self._conversation_service.create_conversation(
            session_id=session_id,
            workspace_id=workspace_id,
            parent_conversation_id=parent_conversation_id,
        )

        return {
            "conversation_id": conversation_id,
            "session_id": session_id,
            "parent_conversation_id": parent_conversation_id,
        }

    async def send_message_to_conversation(
        self,
        conversation_id: str,
        message: str,
        on_complete: Optional[Callable[[Dict[str, Any]], Awaitable[None]]] = None,
    ) -> Dict[str, Any]:
        conversation = self._dao.get_conversation_by_id(conversation_id)
        if not conversation:
            raise ValueError(f"Conversation {conversation_id} not found")

        session = self.get_session(conversation.session_id)
        if not session:
            raise ValueError(f"Session {conversation.session_id} not found")

        async with self._lock:
            if self._conversation_service.is_conversation_running(conversation_id):
                raise RuntimeError(f"Conversation {conversation_id} is already running")

        task = await self._conversation_service.send_user_message(
            conversation_id=conversation_id,
            message=message,
            on_complete=on_complete,
        )

        return {
            "conversation_id": conversation_id,
            "session_id": conversation.session_id,
            "task": task,
        }

    async def end_conversation(self, conversation_id: str) -> int:
        conversation = self._dao.get_conversation_by_id(conversation_id)
        if not conversation:
            return 0

        flushed_count = await self._conversation_service.end_conversation(conversation_id)
        return flushed_count

    async def cancel_conversation(self, conversation_id: str) -> bool:
        conversation = self._dao.get_conversation_by_id(conversation_id)
        if not conversation:
            return False

        result = await self._conversation_service.cancel_conversation(conversation_id)
        return result

    async def delete_conversation(self, conversation_id: str) -> bool:
        conversation = self._dao.get_conversation_by_id(conversation_id)
        if not conversation:
            return False

        return await self._conversation_service.delete_conversation(conversation_id)

    def get_persisted_conversation(self, conversation_id: str) -> Optional[Conversation]:
        return self._dao.get_conversation_by_id(conversation_id)

    async def list_conversation_summaries(self, session_id: int) -> List[Dict[str, Any]]:
        await self._conversation_service.ensure_conversations_loaded(session_id)
        conversations = self._dao.list_conversations_by_session(session_id)
        return [
            {
                "conversation_id": conversation.id,
                "parent_conversation_id": conversation.parent_conversation_id,
                "title": conversation.title,
                "state": conversation.state,
                "message_count": conversation.message_count,
                "created_at": conversation.created_at,
                "updated_at": conversation.updated_at,
            }
            for conversation in conversations
        ]


    async def get_conversation_detail(self, conversation_id: str) -> Optional[Dict[str, Any]]:
        persisted = self._dao.get_conversation_by_id(conversation_id)
        runtime = self._conversation_service.get_state(conversation_id)

        if not persisted and not runtime:
            return None

        if persisted:
            detail = {
                "conversation_id": persisted.id,
                "session_id": persisted.session_id,
                "workspace_id": persisted.workspace_id,
                "parent_conversation_id": persisted.parent_conversation_id,
                "title": persisted.title,
                "state": persisted.state,
                "created_at": persisted.created_at,
                "updated_at": persisted.updated_at,
                "ended_at": persisted.ended_at,
                "message_count": persisted.message_count,
                "error": persisted.error,
            }
        else:
            detail = {
                "conversation_id": runtime["conversation_id"],
                "session_id": runtime["session_id"],
                "parent_conversation_id": runtime.get("parent_conversation_id"),
                "title": runtime.get("title"),
                "state": runtime["state"],
                "created_at": runtime["created_at"],
                "updated_at": runtime["created_at"],
                "ended_at": None,
                "message_count": runtime["message_count"],
                "error": runtime["error"],
            }

        if runtime:
            detail.update({
                "workspace_id": runtime.get("workspace_id"),
                "parent_conversation_id": runtime.get("parent_conversation_id"),
                "title": runtime.get("title"),
                "state": runtime.get("state"),
                "created_at": runtime.get("created_at"),
                "message_count": runtime.get("message_count"),
                "error": runtime.get("error"),
            })

        return detail

    async def get_conversation_messages(self, conversation_id: str) -> List[Dict[str, Any]]:
        # Business message history comes from SQLite nodes plus in-memory ConversationBuffer.
        # It is intentionally not reconstructed from MQ transcripts or conversation-content logs.
        persisted_nodes = self._dao.get_nodes_by_conversation(conversation_id)
        result = [
            {
                "id": node.id,
                "session_id": node.session_id,
                "conversation_id": node.conversation_id,
                "parent_id": node.parent_id,
                "role": node.role,
                "content": node.content,
                "created_at": node.created_at,
            }
            for node in persisted_nodes
        ]

        buffered_nodes = await self._conversation_buffer.get_buffered_nodes(conversation_id)
        if buffered_nodes:
            session_id = await self._conversation_buffer.get_session_id(conversation_id)
            for index, node in enumerate(buffered_nodes):
                result.append({
                    "id": None,
                    "session_id": session_id,
                    "conversation_id": conversation_id,
                    "parent_id": node.parent_id,
                    "role": node.role,
                    "content": node.content,
                    "created_at": node.created_at.isoformat(),
                    "buffer_index": index,
                })

        return result

File: packages/backend/src/service/session-service/session.ts
import { conversationDAO } from '../../data';
import { conversationBuffer } from './conversation-buffer';
import { messageQueue } from './mq';
import { createContentBlock, createMessage, SegmentType } from './canonical';
import { agentService } from '../agent-service/agent';
import type { AgentId } from '../agent-service/adapters';
import { workspaceService } from '../agent-service/service/workspace-service';
import { logger } from '../../core/logging';
import { assertContextMessagesDisjoint, toAgentContextMessages } from './agent-context';

export enum ConversationState {
  PENDING = 'pending',
  RUNNING = 'running',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

interface ConversationInfo {
  conversation_id: string;
  session_id: number;
  workspace_id: string;
  parent_conversation_id: string | null;
  title: string | null;
  state: ConversationState;
  created_at: Date;
  error: string | null;
  message_count: number;
}

export class SessionService {
  private conversations: Map<string, ConversationInfo> = new Map();

  createSession(title: string = '新会话') {
    const userId = 1;
    const workspaceId = this.generateWorkspaceId();
    const sessionId = conversationDAO.createSession(userId, title, workspaceId);
    workspaceService.register(workspaceId, String(sessionId));
    const session = conversationDAO.getSessionById(sessionId);
    if (!session) {
      console.error('[SessionService] getSessionById returned null after createSession, sessionId:', sessionId);
      return {
        id: sessionId,
        user_id: userId,
        title,
        workspace_id: workspaceId,
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
      };
    }
    return session;
  }

  private generateWorkspaceId(): string {
    const timestamp = Date.now();
    const random = Math.random().toString(36).substring(2, 8);
    return `ws-${timestamp}-${random}`;
  }

  deleteSession(sessionId: number): boolean {
    const conversations = conversationDAO.listConversationsBySession(sessionId);
    for (const conv of conversations) {
      this.deleteConversation(conv.id);
    }
    conversationDAO.deleteSession(sessionId);
    return true;
  }

  listSessions() {
    const user = { id: 1 };
    return conversationDAO.listSessionsByUserId(user.id);
  }

  getSession(sessionId: number) {
    return conversationDAO.getSessionById(sessionId);
  }

  async createConversation(
    sessionId: number,
    parentConversationId?: string
  ): Promise<{ conversation_id: string; session_id: number; parent_conversation_id: string | null }> {
    const session = this.getSession(sessionId);
    if (!session) {
      throw new Error(`Session ${sessionId} not found`);
    }

    if (parentConversationId) {
      const parentConv = conversationDAO.getConversationById(parentConversationId);
      if (!parentConv) {
        throw new Error(`Conversation ${parentConversationId} not found`);
      }
      if (parentConv.session_id !== sessionId) {
        throw new Error('Parent conversation does not belong to this session');
      }
    }

    const conversationId = this.generateConversationId();
    const workspaceId = session.workspace_id || conversationId;

    if (session.workspace_id) {
      workspaceService.register(session.workspace_id, String(sessionId));
    }

    conversationDAO.createConversation(
      conversationId,
      sessionId,
      workspaceId,
      ConversationState.PENDING,
      parentConversationId || null
    );

    this.conversations.set(conversationId, {
      conversation_id: conversationId,
      session_id: sessionId,
      workspace_id: workspaceId,
      parent_conversation_id: parentConversationId || null,
      title: null,
      state: ConversationState.PENDING,
      created_at: new Date(),
      error: null,
      message_count: 0,
    });

    return {
      conversation_id: conversationId,
      session_id: sessionId,
      parent_conversation_id: parentConversationId || null,
    };
  }

  async sendMessage(
    conversationId: string,
    message: string,
    enableContext: boolean = false,
    agentId?: AgentId,
    writeConfirmed: boolean = false,
  ): Promise<{ message_id: string; conversation_id: string; session_id: number }> {
    let convInfo = this.conversations.get(conversationId);
    if (!convInfo) {
      const persisted = conversationDAO.getConversationById(conversationId);
      if (!persisted) {
        throw new Error(`Conversation ${conversationId} not found`);
      }
      convInfo = {
        conversation_id: persisted.id,
        session_id: persisted.session_id,
        workspace_id: persisted.workspace_id || conversationId,
        parent_conversation_id: persisted.parent_conversation_id,
        title: persisted.title,
        state: (persisted.state as ConversationState) || ConversationState.PENDING,
        created_at: new Date(persisted.created_at),
        error: persisted.error,
        message_count: persisted.message_count,
      };
      this.conversations.set(conversationId, convInfo);
    }

    if (convInfo.state === ConversationState.RUNNING) {
      throw new Error(`Conversation ${conversationId} is already running`);
    }

    const messageId = this.generateMessageId(conversationId);
    const storedParentChainMessages = enableContext
      ? conversationDAO.getParentChainMessages(conversationId)
      : [];
    const storedCurrentConversationMessages = enableContext
      ? conversationDAO.getMessagesByConversation(conversationId)
      : [];
    assertContextMessagesDisjoint(storedParentChainMessages, storedCurrentConversationMessages);
    const parentChainMessages = toAgentContextMessages(storedParentChainMessages);
    const currentConversationMessages = toAgentContextMessages(storedCurrentConversationMessages);

    await conversationBuffer.createMessage(messageId, conversationId, convInfo.session_id, message);

    convInfo.state = ConversationState.RUNNING;
    convInfo.message_count++;
    conversationDAO.updateConversation(conversationId, {
      state: ConversationState.RUNNING,
      message_count: convInfo.message_count,
      error: null,
    });

    setImmediate(() => {
      agentService.runAgent(
        convInfo!.workspace_id,
        conversationId,
        String(convInfo!.session_id),
        messageId,
        message,
        parentChainMessages,
        currentConversationMessages,
        agentId,
        writeConfirmed,
      ).catch(async (err) => {
        const errorMsg = String(err);
        logger.error({ err: errorMsg, conversationId }, 'Agent run failed');
        try {
          await this.failConversation(conversationId, errorMsg);
        } catch (cleanupErr) {
          logger.error({ err: cleanupErr, conversationId }, 'Failed to mark conversation as failed');
        }
      });
    });

    return {
      message_id: messageId,
      conversation_id: conversationId,
      session_id: convInfo.session_id,
    };
  }

  async endConversation(conversationId: string): Promise<number> {
    const convInfo = this.conversations.get(conversationId);
    if (!convInfo) {
      const persisted = conversationDAO.getConversationById(conversationId);
      if (!persisted) return 0;
    }

    const messages = conversationDAO.getMessagesByConversation(conversationId);
    const actualCount = messages.length;

    if (convInfo) {
      convInfo.state = ConversationState.COMPLETED;
      conversationDAO.updateConversation(conversationId, {
        state: ConversationState.COMPLETED,
        message_count: actualCount,
        ended_at: new Date().toISOString(),
      });
    }

    return actualCount;
  }

  async cancelConversation(conversationId: string): Promise<boolean> {
    logger.info({ conversationId }, 'Cancelling conversation');
    const convInfo = this.conversations.get(conversationId);
    if (!convInfo) {
      const persisted = conversationDAO.getConversationById(conversationId);
      if (!persisted) {
        logger.info({ conversationId }, 'Conversation not found in memory or DB, skip cancel');
        return false;
      }
    }

    if (convInfo) {
      convInfo.state = ConversationState.CANCELLED;
      conversationDAO.updateConversation(conversationId, {
        state: ConversationState.CANCELLED,
        ended_at: new Date().toISOString(),
      });
      logger.info({ conversationId, state: ConversationState.CANCELLED }, 'Conversation state updated');
    }

    try {
      await agentService.cancelConversation(conversationId);
    } catch (cancelErr) {
      logger.error({ err: cancelErr, conversationId }, 'Failed to cancel active agent process');
    }

    await this.finalizeInFlightMessages(conversationId, '运行已取消');
    conversationBuffer.clear(conversationId);
    return true;
  }

  async failConversation(conversationId: string, error: string): Promise<void> {
    const convInfo = this.conversations.get(conversationId);
    if (!convInfo) {
      const persisted = conversationDAO.getConversationById(conversationId);
      if (!persisted) return;
    }

    await this.finalizeInFlightMessages(conversationId, error);

    if (convInfo) {
      convInfo.state = ConversationState.FAILED;
      convInfo.error = error;
      conversationDAO.updateConversation(conversationId, {
        state: ConversationState.FAILED,
        error: error,
        ended_at: new Date().toISOString(),
      });
    }
  }

  private async finalizeInFlightMessages(conversationId: string, error: string): Promise<void> {
    const drafts = conversationBuffer.getDraftsByConversation(conversationId);
    const convInfo = this.conversations.get(conversationId);
    for (const draft of drafts) {
      const baseMeta = { message_id: draft.id };
      await messageQueue.publish(createMessage(
        'assistant',
        draft.id,
        conversationId,
        String(draft.session_id),
        convInfo?.workspace_id ?? draft.conversation_id,
        [createContentBlock(SegmentType.ERROR, error, baseMeta)],
        error
      ));
      await messageQueue.publish(createMessage(
        'assistant',
        draft.id,
        conversationId,
        String(draft.session_id),
        convInfo?.workspace_id ?? draft.conversation_id,
        [createContentBlock(SegmentType.DONE, '', baseMeta)],
        ''
      ));
    }
  }

  async deleteConversation(conversationId: string): Promise<boolean> {
    const persisted = conversationDAO.getConversationById(conversationId);
    if (!persisted && !this.conversations.has(conversationId)) {
      return false;
    }

    conversationDAO.clearChildConversationParents(conversationId);

    for (const [, info] of this.conversations) {
      if (info.parent_conversation_id === conversationId) {
        info.parent_conversation_id = null;
      }
    }

    conversationBuffer.clear(conversationId);
    conversationDAO.deleteMessagesByConversation(conversationId);
    conversationDAO.deleteConversation(conversationId);
    this.conversations.delete(conversationId);

    return true;
  }

  async cascadeDeleteConversation(conversationId: string): Promise<boolean> {
    const persisted = conversationDAO.getConversationById(conversationId);
    if (!persisted && !this.conversations.has(conversationId)) {
      return false;
    }

    const subtreeIds = [conversationId, ...conversationDAO.listDescendantConversationIds(conversationId)];
    let deletedAny = false;

    for (const targetId of subtreeIds.reverse()) {
      deletedAny = (await this.deleteConversation(targetId)) || deletedAny;
    }

    return deletedAny;
  }

  getPersistedConversation(conversationId: string) {
    return conversationDAO.getConversationById(conversationId);
  }

  async updateConversationPositions(
    sessionId: number,
    positions: Array<{ conversation_id: string; x: number; y: number }>
  ): Promise<void> {
    const session = this.getSession(sessionId);
    if (!session) {
      throw new Error(`Session ${sessionId} not found`);
    }

    conversationDAO.updateConversationPositions(sessionId, positions);
  }

  async listConversationSummaries(sessionId: number): Promise<Array<Record<string, unknown>>> {
    const conversations = conversationDAO.listConversationSummariesBySession(sessionId);
    return conversations.map((conv) => ({
      conversation_id: conv.id,
      parent_conversation_id: conv.parent_conversation_id,
      title: conv.title,
      state: conv.state,
      message_count: conv.message_count,
      user_prompt_preview: conv.user_prompt_preview,
      assistant_conclusion_preview: conv.assistant_conclusion_preview,
      created_at: conv.created_at,
      updated_at: conv.updated_at,
      position_x: conv.position_x,
      position_y: conv.position_y,
    }));
  }

  async getConversationDetail(conversationId: string): Promise<Record<string, unknown> | null> {
    const persisted = conversationDAO.getConversationById(conversationId);
    const runtime = this.conversations.get(conversationId);

    if (!persisted && !runtime) return null;

    const messages = conversationDAO.getMessagesByConversation(conversationId);
    const actualMessageCount = messages.length;

    if (persisted) {
      return {
        conversation_id: persisted.id,
        session_id: persisted.session_id,
        workspace_id: persisted.workspace_id,
        parent_conversation_id: persisted.parent_conversation_id,
        title: persisted.title,
        state: persisted.state,
        created_at: persisted.created_at,
        updated_at: persisted.updated_at,
        ended_at: persisted.ended_at,
        message_count: actualMessageCount,
        error: persisted.error,
        position_x: persisted.position_x,
        position_y: persisted.position_y,
      };
    }

    return null;
  }

  async getConversationMessages(conversationId: string): Promise<Array<Record<string, unknown>>> {
    const messages = conversationDAO.getMessagesByConversation(conversationId);
    return messages.map((msg) => ({
      id: msg.id,
      conversation_id: msg.conversation_id,
      session_id: msg.session_id,
      user_content: msg.user_content,
      assistant_content: msg.assistant_content,
      thinking_content: msg.thinking_content,
      content_blocks: msg.content_blocks,
      status: msg.status,
      created_at: msg.created_at,
      updated_at: msg.updated_at,
    }));
  }

  async getParentChainMessages(conversationId: string): Promise<Array<Record<string, unknown>>> {
    const messages = conversationDAO.getParentChainMessages(conversationId);
    return messages.map((msg) => ({
      id: msg.id,
      conversation_id: msg.conversation_id,
      session_id: msg.session_id,
      user_content: msg.user_content,
      assistant_content: msg.assistant_content,
      status: msg.status,
      created_at: msg.created_at,
      updated_at: msg.updated_at,
    }));
  }

  async getContextInfo(conversationId: string): Promise<Record<string, unknown>> {
    const messages = await this.getParentChainMessages(conversationId);
    let totalChars = 0;
    for (const msg of messages) {
      totalChars += (msg.user_content as string)?.length || 0;
      totalChars += (msg.assistant_content as string)?.length || 0;
    }
    const estimatedTokens = Math.floor(totalChars / 4);
    return {
      conversation_id: conversationId,
      message_count: messages.length,
      total_chars: totalChars,
      estimated_tokens: estimatedTokens,
    };
  }

  async recoverStaleConversations(): Promise<number> {
    const staleConvs = conversationDAO.findConversationsByState(ConversationState.RUNNING);
    let recovered = 0;

    for (const conv of staleConvs) {
      logger.info({ conversationId: conv.id, sessionId: conv.session_id }, 'Recovering stale running conversation');
      conversationDAO.updateConversation(conv.id, {
        state: ConversationState.FAILED,
        error: 'Recovered from stale state on startup',
        ended_at: new Date().toISOString(),
      });
      recovered++;
    }

    if (recovered > 0) {
      logger.info({ recovered }, 'Recovered stale running conversations');
    }

    const fixedMessages = conversationDAO.updateMessagesStatusByStatus('streaming', 'failed');
    if (fixedMessages > 0) {
      logger.info({ fixedMessages }, 'Fixed stale streaming messages to failed');
    }

    return recovered;
  }

  private generateMessageId(conversationId: string): string {
    const timestamp = Date.now();
    return `msg-${conversationId}-${timestamp}`;
  }

  private generateConversationId(): string {
    const timestamp = Date.now();
    const random = Math.random().toString(36).substring(2, 8);
    return `conv-${timestamp}-${random}`;
  }
}

export const sessionService = new SessionService();
